File: Sources/MoongateCore/LocalASRConfidence.swift
import Foundation

/// 本地 Whisper 识别输出的置信度概要。Whisper 对中/粤/韩等语言的歌唱内容常**自信地听错**或产出
/// 低置信乱码（青花瓷→「了出情话被风弄转」），此时没有更好的源可换（whisper 本就是 fallback），
/// 诚实做法是给用户「识别质量较低、字幕仅供参考」提示，而非把乱码当成自信字幕呈现。
///
/// **已知局限（保守取舍）**：whisper 置信度是弱信号——部分乱码（如韩语、个别中文）置信度并不低
/// （实测 BLACKPINK avg_prob 0.85 却是乱码）。阈值刻意保守，只对**明显低置信**的输出报警，
/// 零误伤干净内容；代价是 recall 有限（抓不住自信误识）。常量唯一真值在
/// `Tests/fixtures/whisper-timing-constants.json` 的 `localASRConfidence` 段，两端契约断言。
/// 跨端镜像：windows/MoongateCore/LocalAsrConfidence.cs。
public struct LocalASRConfidenceSummary: Codable, Equatable, Sendable {
    public let assessedWordCount: Int
    public let averageProbability: Double
    public let lowConfidenceWordRatio: Double
    public let isLowConfidence: Bool
    public let scriptMismatchRatio: Double
    public let latinTokenRatio: Double
    public let dominantPhraseRatio: Double
    public let repeatedPhraseSpanSeconds: Double
    public let qualityIssues: [String]
    public let isLowQuality: Bool

    public var hasSevereQualityBlocker: Bool {
        qualityIssues.contains("phraseLoop")
            || qualityIssues.contains("autoLanguageMismatch")
    }

    public init(
        assessedWordCount: Int,
        averageProbability: Double,
        lowConfidenceWordRatio: Double,
        isLowConfidence: Bool,
        scriptMismatchRatio: Double = 0,
        latinTokenRatio: Double = 0,
        dominantPhraseRatio: Double = 0,
        repeatedPhraseSpanSeconds: Double = 0,
        qualityIssues: [String] = [],
        isLowQuality: Bool? = nil
    ) {
        self.assessedWordCount = assessedWordCount
        self.averageProbability = averageProbability
        self.lowConfidenceWordRatio = lowConfidenceWordRatio
        self.isLowConfidence = isLowConfidence
        self.scriptMismatchRatio = scriptMismatchRatio
        self.latinTokenRatio = latinTokenRatio
        self.dominantPhraseRatio = dominantPhraseRatio
        self.repeatedPhraseSpanSeconds = repeatedPhraseSpanSeconds
        self.qualityIssues = qualityIssues
        self.isLowQuality = isLowQuality ?? (isLowConfidence || !qualityIssues.isEmpty)
    }

    private enum CodingKeys: String, CodingKey {
        case assessedWordCount
        case averageProbability
        case lowConfidenceWordRatio
        case isLowConfidence
        case scriptMismatchRatio
        case latinTokenRatio
        case dominantPhraseRatio
        case repeatedPhraseSpanSeconds
        case qualityIssues
        case isLowQuality
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let assessedWordCount = try container.decode(Int.self, forKey: .assessedWordCount)
        let averageProbability = try container.decode(Double.self, forKey: .averageProbability)
        let lowConfidenceWordRatio = try container.decode(Double.self, forKey: .lowConfidenceWordRatio)
        let isLowConfidence = try container.decode(Bool.self, forKey: .isLowConfidence)
        let scriptMismatchRatio = try container.decodeIfPresent(Double.self, forKey: .scriptMismatchRatio) ?? 0
        let latinTokenRatio = try container.decodeIfPresent(Double.self, forKey: .latinTokenRatio) ?? 0
        let dominantPhraseRatio = try container.decodeIfPresent(Double.self, forKey: .dominantPhraseRatio) ?? 0
        let repeatedPhraseSpanSeconds = try container.decodeIfPresent(Double.self, forKey: .repeatedPhraseSpanSeconds) ?? 0
        let qualityIssues = try container.decodeIfPresent([String].self, forKey: .qualityIssues) ?? []
        let isLowQuality = try container.decodeIfPresent(Bool.self, forKey: .isLowQuality)
        self.init(
            assessedWordCount: assessedWordCount,
            averageProbability: averageProbability,
            lowConfidenceWordRatio: lowConfidenceWordRatio,
            isLowConfidence: isLowConfidence,
            scriptMismatchRatio: scriptMismatchRatio,
            latinTokenRatio: latinTokenRatio,
            dominantPhraseRatio: dominantPhraseRatio,
            repeatedPhraseSpanSeconds: repeatedPhraseSpanSeconds,
            qualityIssues: qualityIssues,
            isLowQuality: isLowQuality)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(assessedWordCount, forKey: .assessedWordCount)
        try container.encode(averageProbability, forKey: .averageProbability)
        try container.encode(lowConfidenceWordRatio, forKey: .lowConfidenceWordRatio)
        try container.encode(isLowConfidence, forKey: .isLowConfidence)
        try container.encode(scriptMismatchRatio, forKey: .scriptMismatchRatio)
        try container.encode(latinTokenRatio, forKey: .latinTokenRatio)
        try container.encode(dominantPhraseRatio, forKey: .dominantPhraseRatio)
        try container.encode(repeatedPhraseSpanSeconds, forKey: .repeatedPhraseSpanSeconds)
        try container.encode(qualityIssues, forKey: .qualityIssues)
        try container.encode(isLowQuality, forKey: .isLowQuality)
    }
}

public enum LocalASRConfidence {
    /// 平均词概率低于此值视为整体低置信。
    static let averageProbabilityFloor = 0.8
    /// 单词概率低于此值算「低置信词」。
    static let lowConfidenceWordProbability = 0.5
    /// 低置信词占比超过此值视为整体低置信。
    static let lowConfidenceWordRatioCeiling = 0.2
    /// 样本不足时不评估（短片段噪声大），避免误报。
    static let minimumAssessableWordCount = 24

    /// 评估一段转写的整体置信度。只统计有概率、含可见字符的词。
    public static func assess(
        words: [ASRWord],
        segments: [ASRSegment] = [],
        languageCode: String? = nil,
        requestedLanguageCode: String? = nil,
        languageHintCode: String? = nil
    ) -> LocalASRConfidenceSummary {
        var probabilities: [Double] = []
        probabilities.reserveCapacity(words.count)
        for word in words {
            guard word.text.contains(where: { !$0.isWhitespace }) else { continue }
            guard let probability = word.probability else { continue }
            probabilities.append(probability)
        }
        let count = probabilities.count
        let scriptQuality = assessScriptQuality(words: words, languageCode: languageCode)
        let loopIssues = assessLoopQuality(words: words, languageCode: languageCode)
        let segmentQuality = assessSegmentQuality(
            segments: segments,
            languageCode: languageCode,
            requestedLanguageCode: requestedLanguageCode,
            languageHintCode: languageHintCode
        )
        let qualityIssues = Array(Set(scriptQuality.issues + loopIssues + segmentQuality.issues)).sorted()
        guard count > 0 else {
            return LocalASRConfidenceSummary(
                assessedWordCount: 0,
                averageProbability: 1,
                lowConfidenceWordRatio: 0,
                isLowConfidence: false,
                scriptMismatchRatio: scriptQuality.mismatchRatio,
                latinTokenRatio: scriptQuality.latinRatio,
                dominantPhraseRatio: segmentQuality.dominantPhraseRatio,
                repeatedPhraseSpanSeconds: segmentQuality.repeatedPhraseSpanSeconds,
                qualityIssues: qualityIssues)
        }
        let average = probabilities.reduce(0, +) / Double(count)
        let lowRatio = Double(probabilities.filter { $0 < lowConfidenceWordProbability }.count) / Double(count)
        let isLow = count >= minimumAssessableWordCount
            && (average < averageProbabilityFloor || lowRatio > lowConfidenceWordRatioCeiling)
        return LocalASRConfidenceSummary(
            assessedWordCount: count,
            averageProbability: average,
            lowConfidenceWordRatio: lowRatio,
            isLowConfidence: isLow,
            scriptMismatchRatio: scriptQuality.mismatchRatio,
            latinTokenRatio: scriptQuality.latinRatio,
            dominantPhraseRatio: segmentQuality.dominantPhraseRatio,
            repeatedPhraseSpanSeconds: segmentQuality.repeatedPhraseSpanSeconds,
            qualityIssues: qualityIssues)
    }

    public static func assessSubtitle(
        raw: String,
        fileName: String,
        languageCode: String? = nil,
        requestedLanguageCode: String? = nil,
        languageHintCode: String? = nil
    ) -> LocalASRConfidenceSummary {
        let cues = parseSubtitleCues(raw, fileName: fileName)
        let segments = cues.compactMap { cue -> ASRSegment? in
            guard let start = srtTimeToSeconds(cue.start),
                  let end = srtTimeToSeconds(cue.end) else {
                return nil
            }
            return ASRSegment(text: cue.text, startSeconds: start, endSeconds: max(end, start))
        }
        let words = segments.map {
            ASRWord(text: $0.text, startSeconds: $0.startSeconds, endSeconds: $0.endSeconds, probability: nil)
        }
        return assess(
            words: words,
            segments: segments,
            languageCode: languageCode,
            requestedLanguageCode: requestedLanguageCode,
            languageHintCode: languageHintCode
        )
    }

    private static func assessScriptQuality(
        words: [ASRWord],
        languageCode: String?
    ) -> (mismatchRatio: Double, latinRatio: Double, issues: [String]) {
        let language = normalizeLanguageCode(languageCode)
        guard ["ja", "ko", "zh", "yue"].contains(language) else {
            return (0, 0, [])
        }
        var visible = 0
        var expected = 0
        var latin = 0
        for scalar in words.flatMap({ $0.text.unicodeScalars }) {
            if CharacterSet.whitespacesAndNewlines.contains(scalar)
                || CharacterSet.punctuationCharacters.contains(scalar)
                || CharacterSet.symbols.contains(scalar) {
                continue
            }
            visible += 1
            if isExpectedScript(scalar, language: language) {
                expected += 1
            }
            if isLatinLetter(scalar) {
                latin += 1
            }
        }
        guard visible >= minimumAssessableWordCount else {
            return (0, 0, [])
        }
        let expectedRatio = Double(expected) / Double(visible)
        let latinRatio = Double(latin) / Double(visible)
        let mismatchRatio = 1 - expectedRatio
        let issues = expectedRatio < 0.25 && latinRatio > 0.5 ? ["scriptMismatch"] : []
        return (mismatchRatio, latinRatio, issues)
    }

    private static func assessLoopQuality(
        words: [ASRWord],
        languageCode: String?
    ) -> [String] {
        let language = normalizeLanguageCode(languageCode)
        guard ["ja", "ko", "zh", "yue"].contains(language) else { return [] }
        let tokens = words
            .map { normalizeLoopToken($0.text) }
            .filter { !$0.isEmpty }
        guard tokens.count >= minimumAssessableWordCount else { return [] }

        var counts: [String: Int] = [:]
        var maxRun = 0
        var currentRun = 0
        var previous: String?
        for token in tokens {
            counts[token, default: 0] += 1
            if token == previous {
                currentRun += 1
            } else {
                currentRun = 1
                previous = token
            }
            maxRun = max(maxRun, currentRun)
        }
        let uniqueRatio = Double(counts.count) / Double(tokens.count)
        let dominantRatio = Double(counts.values.max() ?? 0) / Double(tokens.count)
        var issues: [String] = []
        if uniqueRatio <= 0.16 {
            issues.append("lowDiversity")
        }
        if maxRun >= 8 || (dominantRatio >= 0.5 && uniqueRatio <= 0.20) {
            issues.append("repetitionLoop")
        }
        return issues
    }

    private static func assessSegmentQuality(
        segments: [ASRSegment],
        languageCode: String?,
        requestedLanguageCode: String?,
        languageHintCode: String?
    ) -> (dominantPhraseRatio: Double, repeatedPhraseSpanSeconds: Double, issues: [String]) {
        let phrases = segments
            .map { segment in
                (
                    phrase: normalizeLoopToken(segment.text),
                    start: segment.startSeconds,
                    end: segment.endSeconds
                )
            }
            .filter { !$0.phrase.isEmpty }

        var issues: [String] = []
        let detected = normalizeLanguageCode(languageCode)
        let hint = normalizeLanguageCode(languageHintCode)
        if isAutoLanguage(requestedLanguageCode)
            && isCJKLanguage(hint)
            && !detected.isEmpty
            && detected != hint {
            issues.append("autoLanguageMismatch")
        }

        guard phrases.count >= 3 else {
            return (0, 0, issues)
        }

        var groups: [String: [(start: Double, end: Double)]] = [:]
        for phrase in phrases {
            groups[phrase.phrase, default: []].append((start: phrase.start, end: phrase.end))
        }
        guard let dominant = groups.max(by: { lhs, rhs in
            lhs.value.count < rhs.value.count
        }) else {
            return (0, 0, issues)
        }
        let dominantCount = dominant.value.count
        let dominantRatio = Double(dominantCount) / Double(phrases.count)
        let earliest = dominant.value.map(\.start).min() ?? 0
        let latest = dominant.value.map(\.end).max() ?? earliest
        var repeatedSpan = max(0, latest - earliest)
        let overallStart = phrases.map(\.start).min() ?? earliest
        let overallEnd = phrases.map(\.end).max() ?? latest
        let overallSpan = max(0, overallEnd - overallStart)

        if groups.count == 1
            && dominantCount >= 3
            && phrases.count <= 4
            && repeatedSpan >= 20 {
            issues.append("lowSegmentDiversity")
        }
        if dominantCount >= 6
            && dominantRatio >= 0.6
            && repeatedSpan >= 25
            && dominant.key.count >= 4 {
            issues.append("phraseLoop")
        }
        if isFragmentedCJKLoop(
            phrases: phrases.map(\.phrase),
            languageCode: detected.isEmpty ? hint : detected,
            groupCount: groups.count,
            repeatedSpan: overallSpan
        ) {
            repeatedSpan = max(repeatedSpan, overallSpan)
            issues.append("lowSegmentDiversity")
            issues.append("phraseLoop")
        }

        return (dominantRatio, repeatedSpan, issues)
    }

    private static func isFragmentedCJKLoop(
        phrases: [String],
        languageCode: String,
        groupCount: Int,
        repeatedSpan: Double
    ) -> Bool {
        guard isCJKLanguage(languageCode),
              phrases.count >= 5,
              groupCount >= 3,
              repeatedSpan >= 20 else {
            return false
        }
        let scalars = phrases.joined().unicodeScalars.filter { scalar in
            !CharacterSet.punctuationCharacters.contains(scalar)
                && !CharacterSet.symbols.contains(scalar)
                && !CharacterSet.whitespacesAndNewlines.contains(scalar)
        }
        guard scalars.count >= 18 else { return false }
        let uniqueRatio = Double(Set(scalars.map(\.value)).count) / Double(scalars.count)
        let averagePhraseLength = Double(scalars.count) / Double(phrases.count)
        return uniqueRatio <= 0.25 && averagePhraseLength <= 8
    }

    private static func normalizeLoopToken(_ raw: String) -> String {
        raw.trimmingCharacters(in: .whitespacesAndNewlines)
            .unicodeScalars
            .filter { scalar in
                !CharacterSet.punctuationCharacters.contains(scalar)
                    && !CharacterSet.symbols.contains(scalar)
            }
            .map(String.init)
            .joined()
            .lowercased()
    }

    private static func normalizeLanguageCode(_ raw: String?) -> String {
        let lower = (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if lower.isEmpty || lower == "auto" || lower == "und" || lower == "unknown" { return "" }
        if lower == "yue" { return "yue" }
        if lower.hasPrefix("zh") { return "zh" }
        if lower.hasPrefix("ja") { return "ja" }
        if lower.hasPrefix("ko") { return "ko" }
        return lower
    }

    private static func isAutoLanguage(_ raw: String?) -> Bool {
        let lower = (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return lower.isEmpty || lower == "auto" || lower == "und" || lower == "unknown"
    }

    private static func isCJKLanguage(_ language: String) -> Bool {
        language == "ja" || language == "ko" || language == "zh" || language == "yue"
    }

    private static func isExpectedScript(_ scalar: UnicodeScalar, language: String) -> Bool {
        switch language {
        case "ko":
            return (0xAC00...0xD7AF).contains(Int(scalar.value))
        case "ja":
            return (0x3040...0x30FF).contains(Int(scalar.value))
                || (0x4E00...0x9FFF).contains(Int(scalar.value))
        case "zh", "yue":
            return (0x4E00...0x9FFF).contains(Int(scalar.value))
        default:
            return false
        }
    }

    private static func isLatinLetter(_ scalar: UnicodeScalar) -> Bool {
        (65...90).contains(Int(scalar.value)) || (97...122).contains(Int(scalar.value))
    }
}

File: windows/MoongateCore/LocalAsrConfidence.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;

namespace Moongate.Core;

/// <summary>
/// Confidence summary of a local Whisper recognition. Whisper often *confidently* mishears or emits
/// low-confidence garbage for sung Chinese/Cantonese/Korean (e.g. 青花瓷 → 「了出情话被风弄转」). There is
/// no better source to switch to (whisper IS the fallback), so the honest behaviour is a "recognition
/// quality is low; for reference only" note rather than presenting garbage as confident subtitles.
///
/// Known limitation (conservative trade-off): whisper confidence is a weak signal — some garbage is
/// confident (BLACKPINK avg_prob 0.85 yet garbled). Thresholds are deliberately conservative: only
/// clearly-low-confidence output is flagged, zero false positives on clean content, at the cost of
/// limited recall. Single source of truth in Tests/fixtures/whisper-timing-constants.json
/// (localASRConfidence); both ends assert their constants equal it. Mirror of Swift LocalASRConfidence.
/// </summary>
public readonly record struct LocalAsrConfidenceSummary
{
    public int AssessedWordCount { get; init; }
    public double AverageProbability { get; init; }
    public double LowConfidenceWordRatio { get; init; }
    public bool IsLowConfidence { get; init; }
    public double ScriptMismatchRatio { get; init; }
    public double LatinTokenRatio { get; init; }
    public double DominantPhraseRatio { get; init; }
    public double RepeatedPhraseSpanSeconds { get; init; }
    public IReadOnlyList<string> QualityIssues { get; init; }
    public bool IsLowQuality { get; init; }
    [JsonIgnore]
    public bool HasSevereQualityBlocker =>
        QualityIssues.Contains("phraseLoop") || QualityIssues.Contains("autoLanguageMismatch");

    public LocalAsrConfidenceSummary(
        int assessedWordCount,
        double averageProbability,
        double lowConfidenceWordRatio,
        bool isLowConfidence,
        double scriptMismatchRatio = 0,
        double latinTokenRatio = 0,
        double dominantPhraseRatio = 0,
        double repeatedPhraseSpanSeconds = 0,
        IReadOnlyList<string>? qualityIssues = null,
        bool? isLowQuality = null)
    {
        AssessedWordCount = assessedWordCount;
        AverageProbability = averageProbability;
        LowConfidenceWordRatio = lowConfidenceWordRatio;
        IsLowConfidence = isLowConfidence;
        ScriptMismatchRatio = scriptMismatchRatio;
        LatinTokenRatio = latinTokenRatio;
        DominantPhraseRatio = dominantPhraseRatio;
        RepeatedPhraseSpanSeconds = repeatedPhraseSpanSeconds;
        QualityIssues = qualityIssues ?? [];
        IsLowQuality = isLowQuality ?? (isLowConfidence || QualityIssues.Count > 0);
    }
}

public static class LocalAsrConfidence
{
    public const double AverageProbabilityFloor = 0.8;
    public const double LowConfidenceWordProbability = 0.5;
    public const double LowConfidenceWordRatioCeiling = 0.2;
    public const int MinimumAssessableWordCount = 24;

    public static LocalAsrConfidenceSummary Assess(
        IReadOnlyList<AsrWord> words,
        string? languageCode = null,
        IReadOnlyList<AsrSegment>? segments = null,
        string? requestedLanguageCode = null,
        string? languageHintCode = null)
    {
        var probabilities = new List<double>(words.Count);
        foreach (var word in words)
        {
            if (string.IsNullOrEmpty(word.Text) || word.Text.All(char.IsWhiteSpace)) continue;
            if (word.Probability is not { } probability) continue;
            probabilities.Add(probability);
        }

        var count = probabilities.Count;
        var scriptQuality = AssessScriptQuality(words, languageCode);
        var loopIssues = AssessLoopQuality(words, languageCode);
        var segmentQuality = AssessSegmentQuality(
            segments ?? [],
            languageCode,
            requestedLanguageCode,
            languageHintCode);
        var qualityIssues = scriptQuality.Issues
            .Concat(loopIssues)
            .Concat(segmentQuality.Issues)
            .Distinct()
            .Order(StringComparer.Ordinal)
            .ToList();
        if (count == 0)
        {
            return new LocalAsrConfidenceSummary(
                0, 1.0, 0.0, false,
                scriptQuality.MismatchRatio,
                scriptQuality.LatinRatio,
                segmentQuality.DominantPhraseRatio,
                segmentQuality.RepeatedPhraseSpanSeconds,
                qualityIssues);
        }

        var average = probabilities.Sum() / count;
        var lowRatio = (double)probabilities.Count(p => p < LowConfidenceWordProbability) / count;
        var isLow = count >= MinimumAssessableWordCount
            && (average < AverageProbabilityFloor || lowRatio > LowConfidenceWordRatioCeiling);
        return new LocalAsrConfidenceSummary(
            count,
            average,
            lowRatio,
            isLow,
            scriptQuality.MismatchRatio,
            scriptQuality.LatinRatio,
            segmentQuality.DominantPhraseRatio,
            segmentQuality.RepeatedPhraseSpanSeconds,
            qualityIssues);
    }

    public static LocalAsrConfidenceSummary AssessSubtitle(
        string raw,
        string fileName,
        string? languageCode = null,
        string? requestedLanguageCode = null,
        string? languageHintCode = null)
    {
        var cues = SrtTools.ParseSubtitle(raw, fileName);
        var segments = cues
            .Select(cue => new
            {
                cue.Text,
                Start = SrtTools.SrtTimeToSeconds(cue.Start),
                End = SrtTools.SrtTimeToSeconds(cue.End),
            })
            .Where(cue => cue.Start is not null && cue.End is not null)
            .Select(cue => new AsrSegment
            {
                Text = cue.Text,
                StartSeconds = cue.Start!.Value,
                EndSeconds = Math.Max(cue.End!.Value, cue.Start.Value),
            })
            .ToList();
        var words = segments
            .Select(segment => new AsrWord
            {
                Text = segment.Text,
                StartSeconds = segment.StartSeconds,
                EndSeconds = segment.EndSeconds,
            })
            .ToList();
        return Assess(words, languageCode, segments, requestedLanguageCode, languageHintCode);
    }

    private static (double MismatchRatio, double LatinRatio, IReadOnlyList<string> Issues) AssessScriptQuality(
        IReadOnlyList<AsrWord> words,
        string? languageCode)
    {
        var language = NormalizeLanguageCode(languageCode);
        if (language is not ("ja" or "ko" or "zh" or "yue"))
        {
            return (0, 0, []);
        }

        var visible = 0;
        var expected = 0;
        var latin = 0;
        foreach (var scalar in string.Concat(words.Select(word => word.Text)).EnumerateRunes())
        {
            if (ShouldIgnoreForScriptQuality(scalar))
            {
                continue;
            }
            visible++;
            if (IsExpectedScript(scalar, language)) expected++;
            if (IsLatinLetter(scalar)) latin++;
        }
        if (visible < MinimumAssessableWordCount)
        {
            return (0, 0, []);
        }

        var expectedRatio = (double)expected / visible;
        var latinRatio = (double)latin / visible;
        var mismatchRatio = 1.0 - expectedRatio;
        IReadOnlyList<string> issues = expectedRatio < 0.25 && latinRatio > 0.5
            ? ["scriptMismatch"]
            : [];
        return (mismatchRatio, latinRatio, issues);
    }

    private static IReadOnlyList<string> AssessLoopQuality(IReadOnlyList<AsrWord> words, string? languageCode)
    {
        var language = NormalizeLanguageCode(languageCode);
        if (language is not ("ja" or "ko" or "zh" or "yue"))
        {
            return [];
        }

        var tokens = words
            .Select(word => NormalizeLoopToken(word.Text))
            .Where(token => token.Length > 0)
            .ToList();
        if (tokens.Count < MinimumAssessableWordCount)
        {
            return [];
        }

        var counts = new Dictionary<string, int>(StringComparer.Ordinal);
        var maxRun = 0;
        var currentRun = 0;
        string? previous = null;
        foreach (var token in tokens)
        {
            counts[token] = counts.GetValueOrDefault(token) + 1;
            if (token == previous)
            {
                currentRun++;
            }
            else
            {
                currentRun = 1;
                previous = token;
            }
            maxRun = Math.Max(maxRun, currentRun);
        }

        var uniqueRatio = (double)counts.Count / tokens.Count;
        var dominantRatio = (double)counts.Values.DefaultIfEmpty(0).Max() / tokens.Count;
        var issues = new List<string>();
        if (uniqueRatio <= 0.16) issues.Add("lowDiversity");
        if (maxRun >= 8 || (dominantRatio >= 0.5 && uniqueRatio <= 0.20)) issues.Add("repetitionLoop");
        return issues;
    }

    private static (double DominantPhraseRatio, double RepeatedPhraseSpanSeconds, IReadOnlyList<string> Issues)
        AssessSegmentQuality(
            IReadOnlyList<AsrSegment> segments,
            string? languageCode,
            string? requestedLanguageCode,
            string? languageHintCode)
    {
        var phrases = segments
            .Select(segment => new
            {
                Phrase = NormalizeLoopToken(segment.Text),
                segment.StartSeconds,
                segment.EndSeconds,
            })
            .Where(segment => segment.Phrase.Length > 0)
            .ToList();
        var issues = new List<string>();
        var detected = NormalizeLanguageCode(languageCode);
        var hint = NormalizeLanguageCode(languageHintCode);
        if (IsAutoLanguage(requestedLanguageCode)
            && IsCjkLanguage(hint)
            && detected.Length > 0
            && detected != hint)
        {
            issues.Add("autoLanguageMismatch");
        }

        if (phrases.Count < 3)
        {
            return (0, 0, issues);
        }

        var groups = phrases
            .GroupBy(segment => segment.Phrase, StringComparer.Ordinal)
            .OrderByDescending(group => group.Count())
            .ToList();
        var dominant = groups.First();
        var dominantCount = dominant.Count();
        var dominantRatio = (double)dominantCount / phrases.Count;
        var earliest = dominant.Min(segment => segment.StartSeconds);
        var latest = dominant.Max(segment => segment.EndSeconds);
        var repeatedSpan = Math.Max(0, latest - earliest);
        var overallSpan = Math.Max(
            0,
            phrases.Max(segment => segment.EndSeconds) - phrases.Min(segment => segment.StartSeconds));

        if (groups.Count == 1
            && dominantCount >= 3
            && phrases.Count <= 4
            && repeatedSpan >= 20)
        {
            issues.Add("lowSegmentDiversity");
        }
        if (dominantCount >= 6
            && dominantRatio >= 0.6
            && repeatedSpan >= 25
            && dominant.Key.Length >= 4)
        {
            issues.Add("phraseLoop");
        }
        if (IsFragmentedCjkLoop(
            phrases.Select(segment => segment.Phrase).ToList(),
            detected.Length == 0 ? hint : detected,
            groups.Count,
            overallSpan))
        {
            repeatedSpan = Math.Max(repeatedSpan, overallSpan);
            issues.Add("lowSegmentDiversity");
            issues.Add("phraseLoop");
        }
        return (dominantRatio, repeatedSpan, issues);
    }

    private static bool IsFragmentedCjkLoop(
        IReadOnlyList<string> phrases,
        string languageCode,
        int groupCount,
        double repeatedSpan)
    {
        if (!IsCjkLanguage(languageCode)
            || phrases.Count < 5
            || groupCount < 3
            || repeatedSpan < 20)
        {
            return false;
        }
        var scalars = string.Concat(phrases)
            .EnumerateRunes()
            .Where(scalar => !ShouldIgnoreForScriptQuality(scalar))
            .ToList();
        if (scalars.Count < 18) return false;
        var uniqueRatio = (double)scalars.Select(scalar => scalar.Value).Distinct().Count() / scalars.Count;
        var averagePhraseLength = (double)scalars.Count / phrases.Count;
        return uniqueRatio <= 0.25 && averagePhraseLength <= 8;
    }

    private static string NormalizeLoopToken(string raw)
    {
        var builder = new StringBuilder();
        foreach (var scalar in (raw ?? "").Trim().EnumerateRunes())
        {
            if (ShouldIgnoreForScriptQuality(scalar)) continue;
            builder.Append(scalar.ToString().ToLowerInvariant());
        }
        return builder.ToString();
    }

    private static string NormalizeLanguageCode(string? raw)
    {
        var lower = (raw ?? "").Trim().ToLowerInvariant();
        if (lower is "" or "auto" or "und" or "unknown") return "";
        if (lower == "yue") return "yue";
        if (lower.StartsWith("zh", StringComparison.Ordinal)) return "zh";
        if (lower.StartsWith("ja", StringComparison.Ordinal)) return "ja";
        if (lower.StartsWith("ko", StringComparison.Ordinal)) return "ko";
        return lower;
    }

    private static bool IsAutoLanguage(string? raw)
    {
        var lower = (raw ?? "").Trim().ToLowerInvariant();
        return lower is "" or "auto" or "und" or "unknown";
    }

    private static bool IsCjkLanguage(string language) => language is "ja" or "ko" or "zh" or "yue";

    private static bool IsExpectedScript(Rune scalar, string language)
    {
        var value = scalar.Value;
        return language switch
        {
            "ko" => value is >= 0xAC00 and <= 0xD7AF,
            "ja" => value is >= 0x3040 and <= 0x30FF or >= 0x4E00 and <= 0x9FFF,
            "zh" or "yue" => value is >= 0x4E00 and <= 0x9FFF,
            _ => false,
        };
    }

    private static bool IsLatinLetter(Rune scalar)
    {
        var value = scalar.Value;
        return value is >= 'A' and <= 'Z' or >= 'a' and <= 'z';
    }

    private static bool ShouldIgnoreForScriptQuality(Rune scalar)
    {
        if (Rune.IsWhiteSpace(scalar)) return true;
        return Rune.GetUnicodeCategory(scalar) switch
        {
            UnicodeCategory.ConnectorPunctuation
                or UnicodeCategory.DashPunctuation
                or UnicodeCategory.OpenPunctuation
                or UnicodeCategory.ClosePunctuation
                or UnicodeCategory.InitialQuotePunctuation
                or UnicodeCategory.FinalQuotePunctuation
                or UnicodeCategory.OtherPunctuation
                or UnicodeCategory.MathSymbol
                or UnicodeCategory.CurrencySymbol
                or UnicodeCategory.ModifierSymbol
                or UnicodeCategory.OtherSymbol => true,
            _ => false,
        };
    }
}

/// <summary>Local-ASR generation result: source SRT path + transcript confidence.</summary>
public readonly record struct GeneratedLocalAsrSource(string Url, LocalAsrConfidenceSummary? Confidence);
